internal/datastore/postgres/readwrite.go

```diff
@@ -53,69 +57,220 @@ type pgReadWriteTXN struct {
+			colNamespace,
+			colObjectID,
+			colRelation,
+			colUsersetNamespace,
+			colUsersetObjectID,
+			colUsersetRelation,
+		))
+
+		sql, args, err := touchInserts.ToSql()
 		if err != nil {
 			return fmt.Errorf(errUnableToWriteRelationships, err)
 		}
 
-		if _, err := rwt.tx.Exec(ctx, sql, args...); err != nil {
+		rows, err := rwt.tx.Query(ctx, sql, args...)
+		if err != nil {
 			return fmt.Errorf(errUnableToWriteRelationships, err)
 		}
+		defer rows.Close()
+
+		// Remove from the TOUCH map of operations each row that was successfully inserted.
+		// This will cover any TOUCH that created an entirely new relationship, acting like
+		// a CREATE.
+		tpl := &core.RelationTuple{
+			ResourceAndRelation: &core.ObjectAndRelation{},
+			Subject:             &core.ObjectAndRelation{},
+		}
+
+		for rows.Next() {
+			err := rows.Scan(
+				&tpl.ResourceAndRelation.Namespace,
+				&tpl.ResourceAndRelation.ObjectId,
+				&tpl.ResourceAndRelation.Relation,
+				&tpl.Subject.Namespace,
+				&tpl.Subject.ObjectId,
+				&tpl.Subject.Relation,
+			)
+			if err != nil {
+				return fmt.Errorf(errUnableToWriteRelationships, err)
+			}
+
+			tplString := tuple.StringWithoutCaveat(tpl)
+			_, ok := touchMutationsByNonCaveat[tplString]
+			if !ok {
+				return spiceerrors.MustBugf("missing expected completed TOUCH mutation")
+			}
+
+			delete(touchMutationsByNonCaveat, tplString)
+		}
+		rows.Close()
+
+		// For each remaining TOUCH mutation, add a DELETE operation for the row iff the caveat and/or
+		// context has changed. For ones in which the caveat name and/or context did not change, there is
+		// no need to replace the row, as it is already present.
+		for _, mut := range touchMutationsByNonCaveat {
+			deleteClauses = append(deleteClauses, exactRelationshipDiffentCaveatClause(mut.Tuple))
+		}
 	}
 
-	if bulkWriteHasValues {
-		sql, args, err := bulkWrite.ToSql()
-		if err != nil {
-			return fmt.Errorf(errUnableToWriteRelationships, err)
+	// Execute the DELETE operation for any DELETE mutations or TOUCH mutations that matched existing
+	// relationships and whose caveat name or context is different in some manner.
+	if len(deleteClauses) == 0 {
+		// Nothing more to do.
+		return nil
+	}
+
+	builder := deleteTuple.Where(deleteClauses)
+
+	// If there are any TOUCH deletions, add RETURNING to track which ones were deleted.
+	if len(touchMutationsByNonCaveat) > 0 {
+		builder = builder.Suffix(fmt.Sprintf("RETURNING %s, %s, %s, %s, %s, %s",
+			colNamespace,
+			colObjectID,
+			colRelation,
+			colUsersetNamespace,
+			colUsersetObjectID,
+			colUsersetRelation,
+		))
+	}
+
+	sql, args, err := builder.
+		Set(colDeletedXid, rwt.newXID).
+		ToSql()
+	if err != nil {
+		return fmt.Errorf(errUnableToWriteRelationships, err)
+	}
+
+	rows, err := rwt.tx.Query(ctx, sql, args...)
+	if err != nil {
+		return fmt.Errorf(errUnableToWriteRelationships, err)
+	}
+	defer rows.Close()
+
+	if len(touchMutationsByNonCaveat) > 0 {
```

Comment:
Would it be safe to just let this run even if it's empty? I think I would prefer that to having huge chunks of code disabled with an `if`.

Reply:
Changed, although I'm sad to leave performance on the table